File: src/main/java/com/github/rogerbarton/intellijsv/SvColorSettingsPage.java
package com.github.rogerbarton.intellijsv;

import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighter;
import com.intellij.openapi.options.colors.AttributesDescriptor;
import com.intellij.openapi.options.colors.ColorDescriptor;
import com.intellij.openapi.options.colors.ColorSettingsPage;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.Map;

public class SvColorSettingsPage implements ColorSettingsPage {
    private static final AttributesDescriptor[] DESCRIPTORS = new AttributesDescriptor[] {
            new AttributesDescriptor("Key", SvSyntaxHighlighter.KEY),
            new AttributesDescriptor("Separator", SvSyntaxHighlighter.SEPARATOR),
            new AttributesDescriptor("Value", SvSyntaxHighlighter.VALUE),
            new AttributesDescriptor("Bad Value", SvSyntaxHighlighter.BAD_CHARACTER),
    };

    @Override
    public @Nullable Icon getIcon() {
        return SvIcons.FILE;
    }

    @Override
    public @NotNull SyntaxHighlighter getHighlighter() {
        return new SvSyntaxHighlighter();
    }

    @Override
    public @NotNull String getDemoText() {
        return "# You are reading the \".properties\" entry.\n" +
                "! The exclamation mark can also mark text as comments.\n" +
                "website = https://en.wikipedia.org/\n" +
                "language = English\n" +
                "# The backslash below tells the application to continue reading\n" +
                "# the value onto the next line.\n" +
                "message = Welcome to \\\n" +
                "          Wikipedia!\n" +
                "# Add spaces to the key\n" +
                "key\\ with\\ spaces = This is the value that could be looked up with the key \"key with spaces\".\n" +
                "# Unicode\n" +
                "tab : \\u0009";
    }

    @Override
    public @Nullable Map<String, TextAttributesKey> getAdditionalHighlightingTagToDescriptorMap() {
        return null;
    }

    @Override
    public AttributesDescriptor @NotNull [] getAttributeDescriptors() {
        return DESCRIPTORS;
    }

    @Override
    public ColorDescriptor @NotNull [] getColorDescriptors() {
        return ColorDescriptor.EMPTY_ARRAY;
    }

    @Override
    public @NotNull String getDisplayName() {
        return "System Verilog";
    }
}

File: src/main/kotlin/com/github/rogerbarton/intellijsv/ide/colors/SvColorSettingsPage.kt
package com.github.rogerbarton.intellijsv.ide.colors

import com.github.rogerbarton.intellijsv.core.SvIcons
import com.github.rogerbarton.intellijsv.core.SvSyntaxHighlighter
import com.intellij.openapi.options.colors.AttributesDescriptor
import com.intellij.openapi.options.colors.ColorDescriptor
import com.intellij.openapi.options.colors.ColorSettingsPage
import com.intellij.openapi.util.io.StreamUtil

class SvColorSettingsPage : ColorSettingsPage
{
    override fun getDisplayName() = "System Verilog"
    override fun getIcon() = SvIcons.SV_FILE
    override fun getAttributeDescriptors(): Array<AttributesDescriptor> = ATTRS
    override fun getColorDescriptors(): Array<ColorDescriptor> = ColorDescriptor.EMPTY_ARRAY
    override fun getHighlighter() = SvSyntaxHighlighter()
    override fun getAdditionalHighlightingTagToDescriptorMap() = ANNOTATOR_TAGS
    override fun getDemoText() = DEMO_TEXT

    companion object
    {
        private val ATTRS = SvColor.values().map { it.attributesDescriptor }.toTypedArray()

        private val ANNOTATOR_TAGS = SvColor.values().associateBy({ it.name }, { it.textAttributesKey })

        private val DEMO_TEXT: String by lazy {
            val stream = SvColorSettingsPage::class.java.classLoader
                .getResourceAsStream("com.github.rogerbarton.intellij.ide/colors/HighlighterDemo.sv")
            StreamUtil.convertSeparators(StreamUtil.readText(stream, "UTF-8"))
        }
    }
}